File: 11_distance_vector/distance_vector.py
# CN Lab Experiment 11: Distance Vector Routing
# Bellman-Ford style updates are used until no routing table changes.

INF = 10**9

n = int(input("Enter number of nodes: "))
print("Enter cost matrix (0 for same node, -1 for no direct link):")
cost = []
for i in range(n):
    row = list(map(int, input().split()))
    for j in range(n):
        if i != j and row[j] == -1:
            row[j] = INF
    cost.append(row)

distance = [row[:] for row in cost]  # Start each node with direct-link costs.

while True:
    changed = False

    for i in range(n):                       # Update table of node i.
        for j in range(n):                   # Try every destination j.
            for k in range(n):               # Try every intermediate neighbor k.
                if distance[i][k] < INF and distance[k][j] < INF:
                    new_cost = distance[i][k] + distance[k][j]
                    if new_cost < distance[i][j]:
                        distance[i][j] = new_cost
                        changed = True

    if not changed:                          # Stop when tables are stable.
        break

for i in range(n):                           # Print each node's routing table.
    print(f"\nRouting table for node {i}:")
    for j in range(n):
        value = "INF" if distance[i][j] >= INF else distance[i][j]
        print(f"Destination {j}: {value}")

File: 11_distance_vector/DistanceVector.java
// CN Lab Experiment 11: Distance Vector Routing
// Repeated Bellman-Ford style updates produce stable routing costs.

import java.util.*;

public class DistanceVector {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        final int INF = 1_000_000_000;

        System.out.print("Enter number of nodes: ");
        int n = scanner.nextInt();

        int[][] distance = new int[n][n]; // Current distance vector tables.

        System.out.println(
            "Enter cost matrix (0 for same node, -1 for no direct link):"
        );

        for (int i = 0; i < n; i++) {           // Read every source row.
            for (int j = 0; j < n; j++) {       // Read every destination cost.
                int value = scanner.nextInt();
                if (i != j && value == -1) {
                    value = INF;                // Mark missing links as infinity.
                }
                distance[i][j] = value;
            }
        }

        boolean changed;

        do {
            changed = false;                    // Assume no table changes.

            for (int i = 0; i < n; i++) {       // Update node i.
                for (int j = 0; j < n; j++) {   // Try destination j.
                    for (int k = 0; k < n; k++) { // Try intermediate k.
                        if (distance[i][k] < INF && distance[k][j] < INF) {
                            int newCost = distance[i][k] + distance[k][j];

                            if (newCost < distance[i][j]) {
                                distance[i][j] = newCost;
                                changed = true; // A better route was found.
                            }
                        }
                    }
                }
            }
        } while (changed);                       // Repeat until stable.

        for (int i = 0; i < n; i++) {            // Print every routing table.
            System.out.println("\nRouting table for node " + i + ":");

            for (int j = 0; j < n; j++) {        // Print each destination.
                String value =
                    distance[i][j] >= INF ? "INF" : String.valueOf(distance[i][j]);
                System.out.println("Destination " + j + ": " + value);
            }
        }

        scanner.close();
    }
}
